motor/core.py

```diff
@@ -1342,60 +1356,29 @@ def to_list(self, length):
             raise pymongo.errors.InvalidOperation(
                 "Can't call to_list on tailable cursor")
 
-        future = self._framework.get_future(self.get_io_loop())
-
-        if not self.alive:
-            future.set_result([])
-        else:
+        async def _coro():
             the_list = []
-            self._framework.add_future(
-                self.get_io_loop(),
-                self._get_more(),
-                self._to_list, length, the_list, future)
-
-        return future
-
-    def _to_list(self, length, the_list, future, get_more_result):
-        # get_more_result is the result of self._get_more().
-        # to_list_future will be the result of the user's to_list() call.
-        try:
-            result = get_more_result.result()
-            # Return early if the task was cancelled.
-            if future.done():
-                return
-            collection = self.collection
-            fix_outgoing = collection.database.delegate._fix_outgoing
-
-            if length is None:
-                n = result
-            else:
-                n = min(length, result)
-
-            for _ in range(n):
-                the_list.append(fix_outgoing(self._data().popleft(),
-                                             collection))
-
-            reached_length = (length is not None and len(the_list) >= length)
-            if reached_length or not self.alive:
-                future.set_result(the_list)
-            else:
-                self._framework.add_future(
-                    self.get_io_loop(),
-                    self._get_more(),
-                    self._to_list, length, the_list, future)
-        except Exception as exc:
-            if not future.done():
-                future.set_exception(exc)
+            if length == 0:
+                return the_list
+            # Else
+            async for doc in self:
+                the_list.append(doc)
+                if (length is not None and
+                        len(the_list) >= length):
+                    break
+            return the_list
+
+        return self._framework.create_task(self.get_io_loop(), _coro())
```

Comment:
I think this pattern is a great example of why we should clean up the codebase to stop returning futures. The code is much more concise, easy to read, and easy to reason about.